### PR TITLE
Renewable + PEM Parameterized Bidder

### PR DESCRIPTION
**Pull request by Xinhe-Chen**
## Fixes
Add the renewable+PEM parameterized bidder (bids the power at a constant price).

## Summary/Motivation:
1. Add a PerfectForecaster to predict the day-ahead and real-time wind power and electricity prices (if needed).
2. Add a PEMParameterizedd bidder to bid the renewable power at a constant prices.

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
3. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/apps/grid_integration/tests/test_perfectforecaster.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/tests/test_perfectforecaster.py:73]: Tests should not return anything - this causes a lot of warnings to be emitted in the test output which makes it hard to find real issues. There are 4 more instances of this below as well.

File: idaes/apps/grid_integration/tests/test_perfectforecaster.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/tests/test_perfectforecaster.py:55]: When checking for errors, you should also include the expected text message to ensure you got the right error (and not another of the same type).

File: idaes/apps/grid_integration/tests/test_perfectforecaster.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/tests/test_perfectforecaster.py:47]: These are more like `component` tests than `unit` tests. Unit tests should generally be focused on individual methods/functions in the code to make sure all the elements work correctly in isolation, followed by a few test cases like these to make sure the end-to-end workflow is correct.

File: idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:23]: Group imports together, starting with native Python packages, then third-party packages, then Pyomo and finally IDAES.

File: idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:28]: You should probably add a check to ensure `cbc` is available in the test environment and to skip these tests if it is not.

File: idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:70]: Commented code.

Also, there is not a lot of testing for this class here - there are a number of methods that should be unit tested.

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
A number of requests regarding improvements to testing.

**Comment by andrewlee94** [comment]
Also, we will need documentation for the new classes and how to use them.

**Comment by Xinhe-Chen** [comment]
> Also, we will need documentation for the new classes and how to use them.

Thank you for your comments! I am still working on improving the tests. The documentation will be added in this PR https://github.com/IDAES/idaes-pse/pull/1090.

**Comment by Xinhe-Chen** [comment]
@radhakrishnatg Hi Radhakrishna, would you please review this PR? Thank you!

**Comment by Xinhe-Chen** [comment]
@adowling2 this is ready for review.

**Review by adowling2 — COMMENTED** [review]
Do the unit tests check all/any error messages thrown (e.g., input validation)?

File: idaes/apps/grid_integration/tests/test_perfectforecaster.py
Comment (andrewlee94) [reply idaes/apps/grid_integration/tests/test_perfectforecaster.py:55]: Tests should have some documentation so the next maintainer knows what they are doing. I try to make the name somewhat indicative, and put inline comments where necessary.

Remember that these are there to remind you when you have to work out why something broke in the future (or help someone else understand).

**Comment by ksbeattie** [comment]
@Xinhe-Chen any news on this?

**Comment by Xinhe-Chen** [comment]
> @Xinhe-Chen any news on this?

I will request a review from Prof. Dowling again.

**Comment by adowling2** [comment]
Tag other people too

File: idaes/apps/grid_integration/bidder.py
Comment (dguittet) [inline idaes/apps/grid_integration/bidder.py:1545]: Typo in renewable

File: idaes/apps/grid_integration/bidder.py
Comment (dguittet) [inline idaes/apps/grid_integration/bidder.py:1574]: typo in grid

File: idaes/apps/grid_integration/bidder.py
Comment (dguittet) [inline idaes/apps/grid_integration/bidder.py:1641]: this could be just "else"

File: idaes/apps/grid_integration/bidder.py
Comment (dguittet) [inline idaes/apps/grid_integration/bidder.py:1651]: shouldn't this use a if-else structure? 
```
if avail_rt_wind == 0:
    bids = [(0, 0), (0, 0)]
else:
    if grid_wind == 0:
        bids = [(0, 0), (avail_rt_wind, self.pem_marginal_cost)]
    else:
        bids = [(0, 0), (grid_wind, 0), (avail_rt_wind, self.pem_marginal_cost)]
```

File: idaes/apps/grid_integration/bidder.py
Comment (dguittet) [inline idaes/apps/grid_integration/bidder.py:1654]: remove print

File: idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py
Comment (dguittet) [inline idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:133]: This test is re-implementing the code/logic of the bidder. While this is possible because the logic is straightforward and simple, it seems to me like the test should use actual numbers for the expected bid curves so that the test pass/fail is not implementation-specific

File: idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py
Comment (dguittet) [inline idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:189]: Same comment as above

Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:188]: Hi Darice, thank you for spending time reviewing the codes! I have addressed all your comments today.

**Review by dguittet — APPROVED** [review]

**Comment by Xinhe-Chen** [comment]
@adowling2 This is ready for review.

**Comment by Xinhe-Chen** [comment]
@ksbeattie Hi Keith, we hope to have this PR reviewed and merged next week.

File: docs/reference_guides/apps/grid_integration/Bidder.rst
Comment (adowling2) [inline docs/reference_guides/apps/grid_integration/Bidder.rst:16]: wind should be lowercase, make sure you define PEM on first use

File: docs/reference_guides/apps/grid_integration/Bidder.rst
Comment (adowling2) [inline docs/reference_guides/apps/grid_integration/Bidder.rst:45]: Equations (4) to (9)

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1406]: What is the specific data structure for a bid? Is this a list? Please provide a little more information in the comments

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1410]: Is this a string or something else?

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1412]: Is this an integer or something else?

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1445]: Please add a few line comments throughout this code to explain what is happening.

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1457]: What does `pair_cnt` stand for? The number of bid elements? Perhaps change to `num_bid_pairs` as the variable name?

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1463]: Does this code pad the bid to ensure it has the correct length? If so, please say this more clearly in the comments.

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1517]: Is this doc string missing arguments (such as bidding_model_object, day_aheady_horizon, ...)?

Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1517]: Is this a float? If so, add (float) or , float to the end of the doc string.

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1519]: Is this a float?

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [inline idaes/apps/grid_integration/bidder.py:1522]: Is this a float?

File: idaes/apps/grid_integration/forecaster.py
Comment (adowling2) [inline idaes/apps/grid_integration/forecaster.py:693]: Perhaps add a sentence explaining the perfect forecast has no error and "sees into the future".

File: idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py
Comment (adowling2) [reply idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:133]: Was this comment addressed? If so, reply here.

**Review by adowling2 — CHANGES_REQUESTED** [review]
Please see the comments below. These are all asking for minor changes, often just more descriptive comments.

File: idaes/apps/grid_integration/bidder.py
Comment (andrewlee94) [reply idaes/apps/grid_integration/bidder.py:1406]: Type hinting would be a good idea. E.g. the method signature could be:

```
def record_bids(self, bids: list, model, date, hour, market):
```

File: idaes/apps/grid_integration/bidder.py
Comment (andrewlee94) [reply idaes/apps/grid_integration/bidder.py:1410]: Typing hinting will help again. Also, remove the line breaks between each argument (otherwise the docs formatter might get confused).

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [reply idaes/apps/grid_integration/bidder.py:1458]: @Xinhe-Chen, did you address this comment? If so, please reply here.

File: idaes/apps/grid_integration/forecaster.py
Comment (adowling2) [reply idaes/apps/grid_integration/forecaster.py:705]: @Xinhe-Chen, did you do this? If so, please reply here.

File: idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py
Comment (adowling2) [reply idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:133]: @Xinhe-Chen Please reply here if you addressed this.

File: idaes/apps/grid_integration/tests/test_perfectforecaster.py
Comment (adowling2) [reply idaes/apps/grid_integration/tests/test_perfectforecaster.py:51]: @Xinhe-Chen, I think you already did this. Is so, just reply here with a short sentence confirming/summarizing what you did.

**Review by adowling2 — COMMENTED** [review]
@Xinhe-Chen, please reply to the open comments to confirm these were addressed.

File: idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:133]: Yes this has been addressed by me.

File: idaes/apps/grid_integration/tests/test_perfectforecaster.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/tests/test_perfectforecaster.py:51]: I have addressed this.

**Comment by Xinhe-Chen** [comment]
> @Xinhe-Chen, please reply to the open comments to confirm these were addressed.

Hi Prof. Dowling, I addressed all comments from you last Friday. 

I am going to to a final check later today to ensure no other issues are remaining later today.

File: docs/reference_guides/apps/grid_integration/Bidder.rst
Comment (Xinhe-Chen) [reply docs/reference_guides/apps/grid_integration/Bidder.rst]: This is addressed.

File: docs/reference_guides/apps/grid_integration/Bidder.rst
Comment (Xinhe-Chen) [reply docs/reference_guides/apps/grid_integration/Bidder.rst]: Done.

File: idaes/apps/grid_integration/bidder.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py]: I added variable types in code comments.

File: idaes/apps/grid_integration/bidder.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py]: Done

File: idaes/apps/grid_integration/bidder.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py]: Done

**Comment by adowling2** [comment]
Excellent. Tag me here when it is ready to review.

File: idaes/apps/grid_integration/bidder.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py]: I explained this at Line 1329. The bid length is correct.

File: idaes/apps/grid_integration/bidder.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py]: Those are inherited from the 'ParameterizedBidder'. I have added the doc strings for them now.

File: idaes/apps/grid_integration/forecaster.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/forecaster.py:705]: Done

**Comment by Xinhe-Chen** [comment]
> Excellent. Tag me here when it is ready to review.

@adowling2 This is ready for review.

**Review by adowling2 — APPROVED** [review]
Ready to merge.

File: idaes/apps/grid_integration/bidder.py
Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py:1454]: Done

File: idaes/apps/grid_integration/bidder.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/bidder.py:1493]: Using a logger is generally preferred to doing `print` statements as this gives the end-user more flexibility in how (and what) things are displayed.

Also, for a new line use `print("\n"}`.

**Review by andrewlee94 — COMMENTED** [review]
@Xinhe-Chen I have one final request before we merge this; someone should be added to the top level CODEOWNERS file as the maintainer for this code (i.e. the person Keith or Ludovico will contact if tests start failing). We require all code in `apps` to have an assigned owner for maintenance purposes.

File: idaes/apps/grid_integration/bidder.py
Comment (andrewlee94) [reply idaes/apps/grid_integration/bidder.py:1493]: https://github.com/IDAES/idaes-pse/blob/main/.github/CODEOWNERS

Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py:1493]: I saw @adowling2 is on the CODEOWNERS roster for the grid_integration codes. I am a PhD student and will graduate in 2 years but I am happy to be responsible for maintaining these codes until I graduate.

Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py]: > Using a logger is generally preferred to doing `print` statements as this gives the end-user more flexibility in how (and what) things are displayed.
> 
> Also, for a new line use `print("\n"}`.

Done.

Comment (andrewlee94) [reply idaes/apps/grid_integration/bidder.py]: I see that you changed the new line, but what about using a logger?

Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py]: I am not very familiar with the logger, so I need to look into that.

Comment (Xinhe-Chen) [reply idaes/apps/grid_integration/bidder.py]: Hi Andrew, I used logger to replace print and it works. Thank you for letting me know this.

**Comment by Xinhe-Chen** [comment]
@andrewlee94 Hi Andrew, I think I have already addressed your comments. Shall we get this PR merged?

File: idaes/apps/grid_integration/bidder.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/bidder.py:762]: You do not need the new line with the logger. Also, in the next line, I would use only a single period (`.`) instead of an elipsis (`...`).

File: idaes/apps/grid_integration/bidder.py
Comment (andrewlee94) [inline idaes/apps/grid_integration/bidder.py:14]: Rather than the base logging package, could you use `import idaes.logger as idaeslog`

`_logger = idaeslog.getLogger(__name__)`

Also note that you do not need to set the logger level here, that is done by the user to control the output.

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
Sorry for the delay, I was at a conference or out on leave the past three weeks. A couple of minor requests for changes to how you are using the logger.

**Comment by Xinhe-Chen** [comment]
> Sorry for the delay, I was at a conference or out on leave the past three weeks. A couple of minor requests for changes to how you are using the logger.

Thank you! I have addressed your new comments.

**Review by andrewlee94 — APPROVED** [review]

**Pull request merged**
